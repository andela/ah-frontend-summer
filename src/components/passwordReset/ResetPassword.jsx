import React from 'react';
import PropTypes from 'prop-types';

import {
    Form, Grid, Segment, Button, Header, Divider, Message
} from 'semantic-ui-react';
import { Link } from 'react-router-dom'

import '../../assets/styles/SignupForm.scss';

const ResetPassword = (props) => {

    const {
        isDisplayed,
        onInputChange,
        disabled,
        loading,
        onSubmit,
        afterSubmit,
        showError,
        errorMsg,
        message
    } = props;

    return (
        <div className="SignupForm">
            <Grid columns={2} relaxed='very' stackable>
                <Grid.Column></Grid.Column>
                <Grid.Column padded='very'>
                    <Segment padded='very'>
                        <Header as='h1' textAlign='center'>
                            Authors Haven
                        </Header>
                        <Header as='h4' color='teal' textAlign='center'>
                            A home for the creative at heart
                        </Header>
                        <br />
                        <br />
                        <div style={{display : !isDisplayed && "none"}}>
                            <Form>
                                <Header as='h5' textAlign='center'>
                                    Enter a new password for your account
                                </Header>
                                { showError ?
                                    <Message color="red">
                                        {errorMsg}
                                    </Message> : null
                                }
                                <div className="spaceComponents">
                                    <Form.Input
                                        type='password'
                                        icon='lock'
                                        iconPosition='left'
                                        placeholder='New Password'
                                        required
                                        onChange={onInputChange}
                                        name="password"
                                    />
                                    <Form.Input
                                        type='password'
                                        icon='lock'
                                        iconPosition='left'
                                        placeholder='Confirm New Password'
                                        required
                                        onChange={onInputChange}
                                        name="confirmPassword"
                                    />
                                    <p style={{color:"gray", fontSize: '10px'}}>
                                    Password must contain only alphabetical letters and numbers</p>
                                    <p style={{color:"gray", fontSize: '10px'}}>
                                    Password must NOT contain special characters</p>
                                    <p style={{color:"gray", fontSize: '10px'}}>
                                    Password must have atleast 8 characters</p>
                                </div>
                                <div className="spaceComponents">
                                    <center>

                                        <Button
                                            color='teal'
                                            size="big"
                                            disabled={disabled}
                                            loading={loading}
                                            onClick={onSubmit}
                                            className="SignupBtn">
                                            Reset Password
                                        </Button>
                                    </center>
                                </div>
                            </Form>
                            <Divider horizontal>Or</Divider>
                            <div className="spaceComponents">
                                <Header color='teal' textAlign='center' padded="very">
                                    <Link to='/'>Home</Link> | <Link to='/login'>Login</Link>
                                </Header>
                            </div>
                        </div>
                        <div style={{display : isDisplayed && "none"}}>
                            <hr/>
                            <br />
                            <br />
                            <Header as='h3' color='teal' textAlign='center'>
                                { message }
                            </Header>
                            <br />
                            <br />
                            <hr/>
                            <br />
                            <div className="spaceComponents">
                            <center>
                                    <Link to="/">
                                        <Button
                                            color='teal'
                                            size="big"
                                            loading={loading}
                                            onClick={afterSubmit}
                                            className="SignupBtn">
                                            OK
                                        </Button>
                                    </Link>
                                </center>
                            </div>
                        </div>
                    </Segment>
                </Grid.Column>
            </Grid>
        </div>
    );
}
ResetPassword.propTypes = {
    isDisplayed: PropTypes.bool.isRequired,
    onInputChange: PropTypes.func.isRequired,
    disabled: PropTypes.bool.isRequired,
    loading: PropTypes.bool.isRequired,
    onSubmit: PropTypes.func.isRequired,
    afterSubmit: PropTypes.func,
    showError: PropTypes.bool.isRequired,
    errorMsg: PropTypes.string,
    message: PropTypes.string
};

ResetPassword.defaultProps = {
    errorMsg: "Error",
    message: "success!"
};

export default ResetPassword;
