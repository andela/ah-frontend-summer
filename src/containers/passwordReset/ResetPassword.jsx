import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import ResetPassword from '../../components/passwordReset/ResetPassword';
import {
    resetPasswordAction
} from '../../store/actions/async/passwordResetActions';
import { resetState } from '../../store/actions/sync/passwordResetActions';
import '../../assets/styles/Signup.scss';

export class PasswordResetContainer extends Component {

    state = {
        buttonDisabled: true,
        password: "",
        confirmPassword: "",
        errorMsg : null
    }

    onInputChange = (e) => {
        this.setState(
            {buttonDisabled : !(this.state.password.length >= 7),
            [e.target.name] : e.target.value}
        );
    }

    onSubmit = () => {
        if (this.state.password !== this.state.confirmPassword){
            this.setState({errorMsg: "Passwords do not match!"})
        }
        else {
            this.setState({errorMsg: null});
            let token = this.props.match.params.token
            const { resetPassword } = this.props;
            resetPassword(token, this.state.password)
        }
    };

    render() {

        const {
            error,
            message,
            loading,
            afterSubmit,
            isDisplayed
        } = this.props;

        const {
            buttonDisabled,
            errorMsg
        } = this.state;

        let showError = false;
        let incomingError = error;

        if (this.state.errorMsg !== null) {
            showError = true
        }else if (incomingError !== null) {
            this.setState({errorMsg: (('user' in incomingError)?
            incomingError.user.message : incomingError.errors.error[0])});
            showError = true
        };

        return (
            <div className="Signup">
                <ResetPassword
                    disabled={buttonDisabled}
                    isDisplayed={isDisplayed}
                    onSubmit={this.onSubmit}
                    onInputChange={this.onInputChange}
                    message={message}
                    afterSubmit={afterSubmit}
                    loading={loading}
                    errorMsg={errorMsg}
                    showError={showError}
                />
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        message: state.passwordReset.message,
        isDisplayed: state.passwordReset.msgDisplayed,
        loading: state.passwordReset.loading,
        error: state.passwordReset.error
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        resetPassword : (token, password) => {
            dispatch(resetPasswordAction(token, password))
        },
        afterSubmit : () => dispatch(resetState())
    };
};

PasswordResetContainer.propTypes = {
    message: PropTypes.string,
    loading: PropTypes.bool,
    isDisplayed: PropTypes.bool.isRequired,
    error: PropTypes.object,
    requestPasswordReset : PropTypes.func,
    afterSubmit : PropTypes.func
};

PasswordResetContainer.defaultProps = {
    loading: false,
    message: "success",
    error:null
};

export default connect(mapStateToProps, mapDispatchToProps)(
    PasswordResetContainer
);

