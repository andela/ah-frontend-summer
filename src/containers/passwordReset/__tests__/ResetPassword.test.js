import React from "react";
import { shallow } from 'enzyme';
import { PasswordResetContainer } from "../ResetPassword";

describe('Reset password container tests', () => {
    it('renders container consistently', () => {
        const Wrapper = shallow(<PasswordResetContainer isDisplayed={false} error={null} />);
        expect(Wrapper).toMatchSnapshot();
    });

    it('renders container correctly with fetch error', () => {
        const Wrapper = shallow(<PasswordResetContainer isDisplayed={true} error={{errors:{error:["error"]}}} />);
        expect(Wrapper).toMatchSnapshot();
    });

    it('renders the ResetPassword Component', () => {
        const Wrapper = shallow(<PasswordResetContainer isDisplayed={false} error={null} />);
        expect(Wrapper.find('ResetPassword')).toHaveLength(1);
    });

    it('renders the main div', () => {
        const Wrapper = shallow(<PasswordResetContainer isDisplayed={false} error={null} />);
        expect(Wrapper.find('div')).toHaveLength(1);
    });

    let wrapper = shallow(<PasswordResetContainer isDisplayed={false} error={null} />);
    let wrapperInstance = wrapper.instance();

    it('throws error when passwords do not match', () => {
        wrapperInstance.setState({
            password: "1234",
            confirmPassword: "12345",
        })
        wrapperInstance.onSubmit();
        expect(wrapperInstance.state.errorMsg).toEqual("Passwords do not match!");
    });

    it('does updates the state on changing input', () => {
        const e = {target: {
            name: "password",
            value: "123456"
        }}
        wrapperInstance.onInputChange(e);
        expect(wrapperInstance.state.password).toEqual("123456");

    });
});
