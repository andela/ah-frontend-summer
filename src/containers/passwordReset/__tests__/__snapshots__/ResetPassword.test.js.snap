// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reset password container tests renders container consistently 1`] = `
<div
  className="Signup"
>
  <ResetPassword
    disabled={true}
    errorMsg={null}
    isDisplayed={false}
    loading={false}
    message="success"
    onInputChange={[Function]}
    onSubmit={[Function]}
    showError={false}
  />
</div>
`;

exports[`Reset password container tests renders container correctly with fetch error 1`] = `
<div
  className="Signup"
>
  <ResetPassword
    disabled={true}
    errorMsg={null}
    isDisplayed={true}
    loading={false}
    message="success"
    onInputChange={[Function]}
    onSubmit={[Function]}
    showError={true}
  />
</div>
`;
